base/Enums.jl

```diff
@@ -84,27 +90,34 @@ macro enum(T,syms...)
             lo = min(lo, i)
             hi = max(hi, i)
         end
+        i += one(i)
     end
-    values = Int32[i[2] for i in vals]
+    values = basetype[i[2] for i in vals]
     if hasexpr && values != unique(values)
         throw(ArgumentError("values for Enum $typename are not unique"))
     end
+    names = Symbol[i[1] for i in vals]
     blk = quote
         # enum definition
-        Base.@__doc__(bitstype 32 $(esc(T)) <: Enum)
+        Base.@__doc__(bitstype $(basetype.size * 8) $(esc(typename)) <: Enum)
+        Base.convert(::Type{Integer}, x::$(esc(typename))) = box($(basetype), x)
+        Base.convert{T<:Integer}(::Type{T}, x::$(esc(typename))) = convert(T, box($(basetype), x))
+        Base.write(io::IO, x::$(esc(typename))) = write(io, $(basetype)(x))
+        Base.read(io::IO, ::Type{$(esc(typename))}) = $(esc(typename))(read(io, $(basetype)))
```

Comment:
I had another go at doing `Enum{T}` and the problems I had last time didn't reappear. So that's nice.

I didn't see a way to implement `Base.read` without a helper method to extract the type, which I think is a bit unfortunate. In principle I _want_ to do

```julia
Base.read{EnumT<:Enum{T<:Integer}}(io::IO, ::Type{EnumT}) = EnumT(read(io, T))
```

but we don't have triangular dispatch yet. :)